qutip/bloch.py

```diff
@@ -394,6 +401,130 @@ def add_annotation(self, state_or_vector, text, **kwargs):
                                  'text': text,
                                  'opts': kwargs})
 
+    def add_arc(self, start, end, fmt="b", steps=None, **kwargs):
+        """Adds an arc between two points on a sphere. The arc is set to be
+        blue solid curve by default.
+
+        The start and end points must be on the same sphere (i.e. have the
+        same radius) but need not be on the unit sphere.
+
+        Parameters
+        ----------
+        start : Qobj or array-like
+            Array with cartesian coordinates of the first point, or a state
+            vector or density matrix that can be mapped to a point on or
+            within the Bloch sphere.
+        end : Qobj or array-like
+            Array with cartesian coordinates of the second point, or a state
+            vector or density matrix that can be mapped to a point on or
+            within the Bloch sphere.
+        fmt : str, default: "b"
+            A matplotlib format string for rendering the arc.
+        steps : int, default: None
+            The number of segments to use when rendering the arc. The default
+            uses 100 steps times the distance between the start and end points,
+            with a minimum of 2 steps.
+        **kwargs : dict
+            Additional parameters to pass to the matplotlib .plot function
+            when rendering this arc.
+        """
+        if isinstance(start, Qobj):
+            pt1 = [
+                expect(sigmax(), start),
+                expect(sigmay(), start),
+                expect(sigmaz(), start),
+            ]
+        else:
+            pt1 = start
+
+        if isinstance(end, Qobj):
+            pt2 = [
+                expect(sigmax(), end),
+                expect(sigmay(), end),
+                expect(sigmaz(), end),
+            ]
+        else:
+            pt2 = end
+
+        pt1 = np.asarray(pt1)
+        pt2 = np.asarray(pt2)
+
+        len1 = np.linalg.norm(pt1)
+        len2 = np.linalg.norm(pt2)
+        if len1 < 1e-12 or len2 < 1e-12:
+            raise ValueError('Polar and azimuthal angles undefined at origin.')
+        elif abs(len1 - len2) > 1e-12:
+            raise ValueError("Points not on the same sphere.")
+        elif (pt1 == pt2).all():
+            raise ValueError(
+                "Start and end represent the same point. No arc can be formed."
+            )
+        elif (pt1 == -pt2).all():
+            raise ValueError(
+                "Start and end are diagonally opposite, no unique arc is"
+                " possible."
+            )
+
+        if steps is None:
+            steps = int(np.linalg.norm(pt1 - pt2) * 100)
+            steps = max(2, steps)
```

Comment:
I still think that dividing by `np.linalg.norm(pt1)` _may_ be useful for the case where the radius of the arc is small. That been said, since I did not check explicitly any example with small arc radius, if you believe that this is enough for most cases lets just leave it as it is and merge.

Reply:
The arc length to chord ratio varies from 1.5 to 1 so I think it's okay. Calculating the arc length and using equal arc steps in the plot is possible a good idea, but also a complete rewrite of the arc algorithm.